lib/consensus.js

```diff
@@ -2211,31 +2083,190 @@ function _computeEndorsement({
+      nonByzantinePeers.push(elector);
+      continue;
     }
-    const supporting = r._proposal._supporting;
-    if(supporting.length > union.length) {
-      union = supporting;
+
+    if(electorTails.length > 1) {
+      // byzantine node, do not tally
+      continue;
     }
-  });
-  return union;
+    const [tail] = electorTails;
+    if(tail._head.length > 1) {
+      // byzantine node, do not tally
+      continue;
+    }
+
+    // non-byzantine peer with head
+    nonByzantinePeers.push(elector);
+    heads.set(elector, tail._head[0]);
+  }
+
+  // go through every peer and determine if they should be prioritized (if
+  // a new merge event from them could progress consensus); if any peer has no
+  // head yet, prioritize it
+  const priorityPeers = new Set(nonByzantinePeers.filter(p => !heads.has(p)));
+  const endorsersKey = recoverySet.size > 0 ?
+    '_recoveryEndorsers' : '_endorsers';
+  for(const [elector, head] of heads) {
+    // skip any peer that is already prioritized
+    if(priorityPeers.has(elector)) {
+      continue;
+    }
+
+    // determine if `elector` needs an endorsement for a decision/recovery
+    // proposal... use this to test whether or not merging each other head
+    // could progress consensus
+    let proposalEndorsers = null;
+    if(head._proposal && !_hasEndorsedProposal(head)) {
+      proposalEndorsers = head._proposal[endorsersKey];
+    } else if(head._recovery && head._recovery.proposal &&
+      !_hasEndorsedRecoveryProposal(head)) {
+      proposalEndorsers = new Set(head._recovery.proposal[endorsersKey]);
+    }
+
+    // check other heads to determine if merging them in would be beneficial
+    for(const [otherElector, otherHead] of heads) {
+      // can't help by merging self
+      if(otherElector === elector) {
+        continue;
+      }
+
+      // if `elector` has a previous MRA from `otherElector`, see if there is
+      // an important change to merge in
+      const mra = head._index && head._index.mostRecentAncestors &&
+        head._index.mostRecentAncestors[otherElector];
+      if(!mra) {
+        // never seen any event from `otherHead`, prioritize `elector` to
+        // merge it in
+        priorityPeers.add(elector);
+        break;
+      }
+
+      if(mra === otherHead) {
+        // `otherHead` is already in elector's ancestry, won't help to merge it
+        continue;
+      }
+
+      // if `otherHead` has new endorsed decision/recovery proposal, then
+      // prioritize to merge it in
+      if((!_hasEndorsedProposal(mra) && _hasEndorsedProposal(otherHead)) ||
+        (!_hasEndorsedRecoveryProposal(mra) &&
+          _hasEndorsedRecoveryProposal(otherHead))) {
+        priorityPeers.add(elector);
+        break;
+      }
+
+      // if `otherHead`'s support has switched, then prioritize `elector` to
+      // merge it in
+      if((!mra && otherHead._supporting) ||
+        (mra && !_compareSupportSet(mra._supporting, otherHead._supporting))) {
```

Comment:
`mra` will always be true here because the `!mra` case has already been handled by the conditional on L2173.